### PR TITLE
Simple syntect-based autoindent

### LINKED ISSUE #424: Auto-indentation
Auto-indentation is one of the major missing features in xi right now. What is our approach?

1. Is providing indentation a plugin's responsibility? (Seems like it.)
1. If providing indentation is a plugin's responsibility, what does the API for that look like?
2. If providing indentation is a plugin's responsibility, how do we approach the syntect plugin? Do we add support upstream (is @trishume open to this?) or do we add support locally?

### PR DESCRIPTION
**Pull request by cmyr**
This adds support for basic autoindent, for languages supported by
syntect, through the syntect-plugin.

This implementation is not quite as fancy as Sublime's; we only use the
two most basic regex patterns, for 'increase next' and 'decrease
current' indentation level.

It is, however, actual honest-to-goodness syntax-aware auto-indent,
which is a step in the right direction.

Closes #424
Progress on #937

-----------

This patch has a sibling at https://github.com/trishume/syntect/pull/223; currently this is pointing to my syntect fork, so it shouldn't be merged before that does. In the meantime, anyone who is interested is encouraged to play around with this; I've only really tested it in rust any python, and not extensively in either case.

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

**Comment by scholtzan** [comment]
The behaviour when writing plain text is slightly weird. If I indent something then it gets automatically unindented. For example, here I am trying to indent a list item: 

out

**Comment by cmyr** [comment]
@scholtzan Okay I've pushed a small change that should prevent us from so aggressively decreasing indent level.

**Comment by scholtzan** [comment]
Cool! It feels much better now :)

**Comment by scholtzan** [comment]
Another thing I noticed is that when text is pasted, it doesn't get indented:
out

File: rust/syntect-plugin/src/main.rs [hunk 1]
```diff
@@ -210,82 +214,171 @@ impl<'a> Syntect<'a> {
...
+            //eprintln!("auto indenting {}, prev_level {}", line, base_indent);
+            self.set_indent(view, line, indent_level)
+        } else {
+            Ok(())
+        }
+    }
 
-        if is_newline {
-            let line_num = view.line_of_offset(start).unwrap();
+    /// Called when actviely editing a line; cheifly checks for whether or not
```
Comment (scholtzan) [inline rust/syntect-plugin/src/main.rs]: Nit: actively; chiefly

Comment (scholtzan) [reply rust/syntect-plugin/src/main.rs]: But this feels pretty good when writing code :)

**Comment by cmyr** [comment]
Oups, missed that paste comment @scholtzan. I think that's a slightly different feature, and is out of scope for this PR; that and bulk-reindent are things that I'll open issues for afterwards.

File: rust/syntect-plugin/src/main.rs [hunk 2]
```diff
@@ -210,82 +219,175 @@ impl<'a> Syntect<'a> {
...
+    fn autoindent_line(&mut self, view: &mut MyView, line: usize) -> Result<(), Error> {
+        let _t = trace_block("Syntect::autoindent", &["syntect"]);
+        debug_assert!(line > 0);
+        let tab_size = view.get_config().tab_size;
+        let current_indent = self.indent_level_of_line(view, line);
+        let base_indent = self
+            .previous_nonblank_line(view, line)?
+            .map(|l| self.indent_level_of_line(view, l))
+            .unwrap_or(0);
+        let increase_level = self.test_increase(view, line)?;
```
Comment (scholtzan) [inline rust/syntect-plugin/src/main.rs]: If I test here whether it should be decreased with `test_decrease` it returns an error (`error in autoindent PeerDisconnect`). The reason for this seems to be that at some point in `get_metadata` were are trying to get the `line` with `view.get(line)`. View seems to access some cache with the lines but cannot find `line` and returns an error. I am not sure if that is the expected behaviour?

**Pull request merged**